File: Question 1475 - Final Prices With a Special Discount in a Shop/src/Solution.java
import java.util.Stack;

class Solution {
    public int[] finalPrices(int[] prices) {
        int n= prices.length;;
        int[] ans = new int[n];

        Stack<Integer> s = new Stack<>();
        s.push(0);

        for(int i=n-1; i>= 0; i--)
        {
            int curr = prices[i];
            while (s.peek() > curr)
                s.pop();

            ans[i] = prices[i] - s.peek();
            s.push(curr);
        }
        return ans;
    }
}

File: Question 1475 - Final Prices With a Special Discount in a Shop/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> finalPrices(vector<int> &prices)
    {
        vector<int> ans;
        stack<int> st;

        for (int i = prices.size() - 1; i >= 0; i--)
        {
            while (!st.empty() && st.top() > prices[i])
                st.pop();

            int disc = st.empty() ? 0 : st.top();

            ans[i] = prices[i] - disc;

            st.push(prices[i]);
        }

        return ans;
    }
};